File: cardmasters_app/cardmasters_app/report/work_order_operations/work_order_operations.py
"""Independent Work Order lists grouped by operation."""

import frappe
from frappe import _


SOURCE_FIELDS = [
    "sales_order", "material_request", "custom_sales_return_reference",
    "custom_document", "custom_document_id",
]


def source_order(work_order):
    # A return can retain its original Sales Order; the return is the source.
    if work_order.get("custom_sales_return_reference"):
        return "Delivery Note", work_order["custom_sales_return_reference"]
    if work_order.get("material_request"):
        return "Material Request", work_order["material_request"]
    if work_order.get("custom_document") in ("Sales Order", "Material Request"):
        if work_order.get("custom_document_id"):
            return work_order["custom_document"], work_order["custom_document_id"]
    if work_order.get("sales_order"):
        return "Sales Order", work_order["sales_order"]
    return "", ""


def get_source_labels(work_orders):
    sources = {}
    for work_order in work_orders:
        source_type, source_name = source_order(work_order)
        if source_type in ("Sales Order", "Material Request"):
            sources.setdefault(source_type, set()).add(source_name)
    labels = {}
    for source_type, names in sources.items():
        if not frappe.has_permission(source_type, "read"):
            continue
        field = "customer_name" if source_type == "Sales Order" else "company"
        names = sorted(names)
        for start in range(0, len(names), 500):
            for source in frappe.get_list(
                source_type, filters={"name": ["in", names[start:start + 500]]},
                fields=["name", field], limit_page_length=0,
            ):
                labels[(source_type, source.name)] = source.get(field) or ""
    return labels


def build_result(work_orders, operations, source_labels=None):
    source_labels = source_labels or {}
    orders = {row["name"]: row for row in work_orders}
    groups = {}
    for operation in operations:
        if operation["parent"] in orders and operation.get("operation"):
            groups.setdefault(operation["operation"], []).append(operation)

    columns, data = [], []
    for index, (operation_name, entries) in enumerate(sorted(groups.items())):
        prefix = f"operation_{index}"
        for suffix, label, fieldtype, options, width in (
            ("source_order", _("Source Order"), "Dynamic Link", f"{prefix}_source_type", 340),
            ("work_order", _("Work Order"), "Link", "Work Order", 210),
            ("item_name", _("Item Name"), "Data", None, 200),
            ("particulars", _("Particulars"), "Text", None, 300),
            ("progress", _("Progress"), "Data", None, 130),
        ):
            columns.append({
                "fieldname": f"{prefix}_{suffix}", "label": label,
                "fieldtype": fieldtype, "options": options, "width": width,
                "operation_group": operation_name, "sortable": False, "dropdown": False,
            })
        columns.append({
            "fieldname": f"{prefix}_source_type", "label": _("Source Type"),
            "fieldtype": "Data", "hidden": 1,
        })
        for row_index, entry in enumerate(entries):
            if row_index == len(data):
                data.append({})
            work_order = orders[entry["parent"]]
            source_type, source_name = source_order(work_order)
            data[row_index].update({
                f"{prefix}_work_order": entry["parent"],
                f"{prefix}_item_name": work_order.get("item_name") or "",
                f"{prefix}_particulars": work_order.get("custom_particulars") or "",
                f"{prefix}_source_type": source_type,
                f"{prefix}_source_order": source_name,
                f"{prefix}_source_label": source_labels.get((source_type, source_name), ""),
                f"{prefix}_progress": entry.get("custom_progress") or "",
            })
    return columns, data


def execute(filters=None):
    filters = frappe._dict(filters or {})
    conditions = [["Work Order", "docstatus", "<", 2]]
    for field in ("company", "name", "status"):
        value = filters.get("work_order" if field == "name" else field)
        if value:
            conditions.append(["Work Order", field, "=", value])
    if filters.branch:
        branch_field = (
            "custom_production_branch"
            if filters.branch_source == "custom_production_branch"
            else "custom_for_branch"
        )
        conditions.append(["Work Order", branch_field, "=", filters.branch])
    meta = frappe.get_meta("Work Order")
    fields = ["name", "item_name"] + [
        field for field in [*SOURCE_FIELDS, "custom_particulars"] if meta.has_field(field)
    ]
    if not frappe.get_meta("Work Order Operation").has_field("custom_progress"):
        frappe.throw(_("Work Order Operation requires the custom_progress field."))
    operation_filters = {"operation": ["is", "set"]}
    for field, value in (("operation", filters.operation), ("custom_progress", filters.progress)):
        if value:
            # Preserve single-value URLs/API calls alongside multi-select filters.
            operation_filters[field] = ["in", value] if isinstance(value, list) else value
    for field, value in operation_filters.items():
        operator, operand = value if isinstance(value, list) else ("=", value)
        conditions.append(["Work Order Operation", field, operator, operand])
    # Apply Work Order permissions before reading any of its child rows.
    # Joining matching operations avoids loading thousands of unrelated orders.
    orders = frappe.get_list(
        "Work Order", filters=conditions, fields=fields, distinct=True,
        order_by="`tabWork Order`.creation asc, `tabWork Order`.name asc", limit_page_length=0,
    )
    if not orders:
        return [], []
    operations = []
    names = [row.name for row in orders]
    for start in range(0, len(names), 500):
        child_filters = {
            "parent": ["in", names[start:start + 500]],
            "parenttype": "Work Order", "parentfield": "operations",
            **operation_filters,
        }
        operations.extend(frappe.get_all(
            "Work Order Operation", filters=child_filters,
            fields=["parent", "operation", "custom_progress", "idx"],
            order_by="parent asc, idx asc",
        ))
    order_position = {name: index for index, name in enumerate(names)}
    operations.sort(key=lambda row: (order_position[row.parent], row.idx))
    return build_result(orders, operations, get_source_labels(orders))

File: cardmasters_app/cardmasters_app/report/work_order_operations/work_order_operations.js
frappe.query_reports["Work Order Operations"] = {
    filters: [
        { fieldname: "company", label: __("Company"), fieldtype: "Link", options: "Company",
            default: frappe.defaults.get_user_default("Company") },
        { fieldname: "work_order", label: __("Work Order"), fieldtype: "Link", options: "Work Order" },
        { fieldname: "branch_source", label: __("Branch Source"), fieldtype: "Select",
            options: [
                { value: "custom_for_branch", label: __("For Branch") },
                { value: "custom_production_branch", label: __("Production Branch") },
            ], default: "custom_for_branch" },
        { fieldname: "branch", label: __("Branch"), fieldtype: "Link", options: "Branch" },
        { fieldname: "operation", label: __("Operations"), fieldtype: "MultiSelectList",
            options: "Operation",
            get_data(txt) { return frappe.db.get_link_options("Operation", txt); } },
        { fieldname: "progress", label: __("Progress"), fieldtype: "MultiSelectList",
            get_data(txt) {
                return ["Not Started", "In Progress", "On Hold", "Done"]
                    .filter(value => __(value).toLowerCase().includes((txt || "").toLowerCase()))
                    .map(value => ({ value, description: __(value) }));
            } },
        { fieldname: "status", label: __("Work Order Status"), fieldtype: "Select",
            options: "\nDraft\nNot Started\nIn Process\nCompleted\nStopped\nClosed" },
    ],

    onload(report) {
        // Let several quick multi-select clicks settle before requesting the grid.
        const refresh = frappe.utils.debounce(() => {
            if (!report._no_refresh && report.report_name === "Work Order Operations") {
                report.refresh(true);
            }
        }, 350);
        report.filters.filter(filter => ["operation", "progress"].includes(filter.fieldname))
            .forEach(filter => {
                filter.on_change = () => {
                    if (!report._no_refresh) refresh();
                };
            });
    },

    formatter(value, row, column, data, default_formatter) {
        const formatted = default_formatter(value, row, column, data);
        if (!column.fieldname?.endsWith("_source_order")) return formatted;
        const label = data?.[column.fieldname.replace(/_source_order$/, "_source_label")];
        return label ? `${formatted} · ${frappe.utils.escape_html(label)}` : formatted;
    },

    get_datatable_options(options) {
        return Object.assign(options, {
            serialNoColumn: false, checkboxColumn: false,
            disableReorderColumn: true, inlineFilters: false,
        });
    },

    after_datatable_render(datatable) {
        // Keep the real grid (and its link formatting, resizing and export data).
        // The extra band lives inside the header, so native horizontal scrolling
        // moves both header levels together without a second scroll handler.
        datatable.cmOperationGroupsCleanup?.();
        const header = datatable.header;
        const band = document.createElement("div");
        band.className = "cm-operation-groups";
        band.style.cssText = "display:flex;order:-1;width:max-content";
        header.style.display = "flex";
        header.style.flexDirection = "column";
        header.appendChild(band);

        // Scope real borders to this grid, including virtualized rows and headers.
        // A shared rule keeps the upper band and group-ending columns identical.
        const separators = document.createElement("style");
        const columns = datatable.datamanager.getColumns();
        const scope = `.${datatable.style.scopeClass}`;
        const groupEnds = columns.flatMap((column, index) => {
            if (!column.operation_group ||
                column.operation_group === columns[index + 1]?.operation_group) return [];
            return [`${scope} .dt-cell--col-${column.colIndex}`];
        });
        separators.textContent = `
            ${scope} .dt-cell {
                border-right: 1px solid var(--border-color, #d1d8dd);
            }
            ${scope} .cm-operation-group {
                flex: none;
                box-sizing: border-box;
                padding: 8px;
                text-align: center;
                font-weight: 600;
                overflow: hidden;
                text-overflow: ellipsis;
                white-space: nowrap;
                border-top: 1px solid var(--border-color, #d1d8dd);
                background: var(--subtle-fg, var(--fg-color, #f5f7fa));
            }
            ${scope} .cm-operation-group:first-child {
                border-left: 1px solid var(--border-color, #d1d8dd);
            }
            ${[`${scope} .cm-operation-group`, ...groupEnds].join(",")} {
                border-right: 2px solid var(--text-light, #98a2b3);
            }
        `;
        datatable.wrapper.appendChild(separators);

        let frame = null;
        let disposed = false;
        let groupCells = [];
        const sync = () => {
            frame = null;
            if (disposed) return;
            const columns = datatable.datamanager.getColumns();
            const cells = header.querySelectorAll(".dt-row-header > .dt-cell");
            const groups = [];
            cells.forEach((cell, index) => {
                const name = columns[index]?.operation_group || "";
                let group = groups[groups.length - 1];
                if (!group || group.name !== name) {
                    group = { name, width: 0 };
                    groups.push(group);
                }
                group.width += cell.getBoundingClientRect().width;
            });
            const structureChanged = groups.length !== groupCells.length ||
                groups.some((group, index) => group.name !== groupCells[index]?.name);
            if (structureChanged) {
                groupCells = groups.map(group => {
                    const cell = document.createElement("div");
                    cell.className = "cm-operation-group";
                    cell.textContent = group.name;
                    cell.title = group.name;
                    cell.style.width = `${group.width}px`;
                    return { name: group.name, width: group.width, cell };
                });
                band.replaceChildren(...groupCells.map(group => group.cell));
            } else {
                groups.forEach((group, index) => {
                    if (group.width !== groupCells[index].width) {
                        groupCells[index].cell.style.width = `${group.width}px`;
                        groupCells[index].width = group.width;
                    }
                });
            }
        };
        // The grid sets every column width during refresh. Measuring the entire
        // header after each write forces repeated browser layouts on wide reports.
        const scheduleSync = () => {
            if (!disposed && frame === null) frame = requestAnimationFrame(sync);
        };
        const observer = new ResizeObserver(scheduleSync);
        header.querySelectorAll(".dt-row-header > .dt-cell").forEach(cell => observer.observe(cell));
        // Schedule during drag too; observer delivery can lag behind pointer updates.
        const manager = datatable.columnmanager;
        const setHeaderWidth = manager.setColumnHeaderWidth;
        manager.setColumnHeaderWidth = function(...args) {
            const result = setHeaderWidth.apply(this, args);
            scheduleSync();
            return result;
        };
        scheduleSync();
        const cleanup = () => {
            disposed = true;
            if (frame !== null) cancelAnimationFrame(frame);
            observer.disconnect();
            manager.setColumnHeaderWidth = setHeaderWidth;
            band.remove();
            separators.remove();
        };
        datatable.cmOperationGroupsCleanup = cleanup;
        if (!datatable.cmOperationGroupsDestroyBound) {
            datatable.on("onDestroy", () => datatable.cmOperationGroupsCleanup?.());
            datatable.cmOperationGroupsDestroyBound = true;
        }
    },
};
